### ISSUE TITLE
[QUESTION] How to use the trained model in tensorflow?

### ISSUE THREAD

**Issue by nckenn** [Horea94/Fruit-Images-Dataset#22]
Just want to know, is there any example how to use the generated model.h5 in tensorflow to classify image?

**Comment by Horea94** [comment]
In the [Fruits-360 CNN.py](src/image_classification/Fruits-360 CNN.py) file, we are using 
`model.load_weights(model_out_dir + "/model.h5")`
to load the saved weights from the .h5 file. Then, using
`y_pred = model.predict(testGen, steps=(testGen.n // batch_size) + 1, verbose=verbose)`
we store the prediction in the y_pred variable. The variable is a list of arrays. Each array contains a probability for each of the classes, the index of the highest probability being the predicted class. To get the class name, you can look at the same index in the labels array.
The predict method is not limited to generators. According to the documentation found here [https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict](https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict), the input can be:
- A Numpy array (or array-like), or a list of arrays (in case the model has multiple inputs).
- A TensorFlow tensor, or a list of tensors (in case the model has multiple inputs).
- A tf.data dataset.
- A generator or keras.utils.Sequence instance. A more detailed description of unpacking behaviour for iterator types (Dataset, generator, Sequence) is given in the Unpacking behaviour for iterator-like inputs section of Model.fit.

**Comment by nckenn** [comment]
> In the [Fruits-360 CNN.py](src/image_classification/Fruits-360 CNN.py) file, we are using
> `model.load_weights(model_out_dir + "/model.h5")`
> to load the saved weights from the .h5 file. Then, using
> `y_pred = model.predict(testGen, steps=(testGen.n // batch_size) + 1, verbose=verbose)`
> we store the prediction in the y_pred variable. The variable is a list of arrays. Each array contains a probability for each of the classes, the index of the highest probability being the predicted class. To get the class name, you can look at the same index in the labels array.
> The predict method is not limited to generators. According to the documentation found here https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict, the input can be:
> 
> * A Numpy array (or array-like), or a list of arrays (in case the model has multiple inputs).
> * A TensorFlow tensor, or a list of tensors (in case the model has multiple inputs).
> * A tf.data dataset.
> * A generator or keras.utils.Sequence instance. A more detailed description of unpacking behaviour for iterator types (Dataset, generator, Sequence) is given in the Unpacking behaviour for iterator-like inputs section of Model.fit.

Thank you for quick reply, very much appreciated, can you take a look at my code below, im still new ML, thats why i dont understand some of the terms you used. Am i in the right path?

`    image = cv.imread('assets/img/banana.jpg')
    image = cv.resize(image, (100, 100))

    model = tensorflow.keras.models.load_weights('models/model.h5', custom_objects={'tf' : tf})
    data = np.ndarray(shape=(1, 100, 100, 3), dtype=np.float32)
    
    image_array = np.asarray(image)
    normalized_image_array = (image_array.astype(np.float32) / 127.0) - 1
    data[0] = normalized_image_array
    predictions = model.predict(data,1)
    
    classes = predictions.argmax(axis=1)
    print(predictions)
`

**Comment by Horea94** [comment]
The overall logic seems fine. However, in our model, we are representing images as integers rather than floats and we do not apply normalisation. So maybe you could change data to:
`data = np.ndarray(shape=(1, 100, 100, 3), dtype=np.float32)`
and use the image_array instead of the normalized_image_array.

**Comment by nckenn** [comment]
> The overall logic seems fine. However, in our model, we are representing images as integers rather than floats and we do not apply normalisation. So maybe you could change data to:
> `data = np.ndarray(shape=(1, 100, 100, 3), dtype=np.float32)`
> and use the image_array instead of the normalized_image_array.

Im still confused :(, how can i classify the label? there is no label.txt generated after running Fruits-360 CNN.py, do i need to create manually?

**Comment by Horea94** [comment]
Fruits-360 CNN.py uses an in-memory array of labels. By default, this array is built as the list of all the folder names in the Training folder. If the use_label_file is set to true, it will look for a labels.txt file that the user creates. That file should contain the label names (folder names from the Training folder) that you want to train the model for, one per line.
So when predicting with the trained model, you can build the labels array with similar logic, and use it to convert the predicted class(which will be a number) to a human-readable name.

**Comment by nckenn** [comment]
> Fruits-360 CNN.py uses an in-memory array of labels. By default, this array is built as the list of all the folder names in the Training folder. If the use_label_file is set to true, it will look for a labels.txt file that the user creates. That file should contain the label names (folder names from the Training folder) that you want to train the model for, one per line.
> So when predicting with the trained model, you can build the labels array with similar logic, and use it to convert the predicted class(which will be a number) to a human-readable name.

now everything works fine, except for this one issue, when i use the image from the Training folder as input image, the prediction is accurate, but when i used the image downloaded from google, the prediction is not correct? any suggestion on my code?

**Comment by Horea94** [comment]
The issue is most likely not in the code.
The images in the dataset are taken under the same lighting conditions, using one fruit per class.
For the model to properly generalise on real-world images, you would need a lot more variance in the data.
Some degree of variance can be introduced with data augmentation(image flips, hue/saturation/light changes). Perhaps use image processing to alter the background of the training images, as the dataset only uses a white background.
Alternatively, you could add more images to your training dataset.
